core/capabilities/webapi/outgoing_connector_handler.go

```diff
@@ -153,6 +146,69 @@ func (c *OutgoingConnectorHandler) HandleSingleNodeRequest(ctx context.Context,
 	}
 }
 
+// AwaitConnection attempts to establish a connection to an available gateway.  It iterates through available gateways
+// using the gatewaySelector, attempting to connect to each.  The method respects the provided context, allowing for
+// cancellation or timeout.
+func (c *OutgoingConnectorHandler) AwaitConnection(ctx context.Context) (string, error) {
+	return c.awaitConnectionUntilCanceled(ctx)
+}
+
+func (c *OutgoingConnectorHandler) awaitConnectionUntilCanceled(ctx context.Context) (string, error) {
+	attempts := make(map[string]int)
+	wait := 10 * time.Millisecond
+	for {
+		select {
+		case <-ctx.Done():
+			return "", ctx.Err()
+		default:
+			gateway, err := c.gatewaySelector.NextGateway()
+			if err != nil {
+				return "", fmt.Errorf("failed to select gateway: %w", err)
+			}
+
+			// if called before, we have tried them all, so backoff
+			if attempts[gateway] > 0 {
+				c.lggr.Warnw("all available gateway nodes attempted without connection, backing off", "waitTime", wait)
+				attempts = make(map[string]int)
+				<-time.After(wait)
+				wait *= 2
```

Comment:
Is an exponential backoff overkill here? I'm thinking how this can go wrong.

Imagine: one of two gateways is down and workflow execution load is high. `c.gatewaySelector.NextGateway()` will be called quite often. It's a coin flip which one it selects. There is a chance that it keeps selecting the down gateway over and over without choosing to the healthy one.

The above scenario is baked into the design because we don't have a connection manager (unhealthy gateways will still be selected).

Exponential backoff seems like it might cause this scenario to happen more severely. This is an edge case, but could have impact if encountered on time sensitive workflows.

Reply:
discussed: resolution is to create a new round robin selector per request instead of a shared selector